src/IO/ReadBufferFromS3.cpp

```diff
@@ -152,11 +152,12 @@ bool ReadBufferFromS3::nextImpl()
         }
     }
 
+    /// use the buffer returned by `impl`
+    BufferBase::set(impl->buffer().begin(), impl->buffer().size(), impl->offset());
+
     if (!next_result)
         return false;
 
-    BufferBase::set(impl->buffer().begin(), impl->buffer().size(), impl->offset()); /// use the buffer returned by `impl`
```

Comment:
I do not see how this should fix failed assertion. Why `hasPendingData` will become `true` if we reached end of file and so size() and offset() should be 0?

Reply:
In case of empty file `BufferBase::set()` will never be called.
So when it first enters `ReadBufferFromS3::nextImpl()` everything is correct it returns false.
However when it will be called next time, it will enter here https://github.com/ClickHouse/ClickHouse/blob/0b40e05ffcb9e5986425da0d6f61969894168017/src/IO/ReadBufferFromS3.cpp#L98-L99

Before those lines `impl->position()` is `impl->working_buffer.end_pos`,
after assignment `impl->position()` will be corrupted, since it will `position()` is just NULL,
and so `impl->hasPendingData()` (which is `available() > 0`), will return some huge number (non-null pointer - NULL)